Adding internal documentation about requirements for object stores (#147928)
elastic/elasticsearch: docs/internal/object-store-requirements.md

> # **Elasticsearch Requirements for Object Store**

Reviewer: ```suggestion
# **Elasticsearch Requirements for a S3-compatible Object Store**
```

Also lets rename the file to `s3-object-store-requirements.md`.

Just to avoid confusion that GCS or Azure has anything to do with this document (which is heavily focused on S3 APIs).

> 🟢 = tested in `/_snapshot/{repository}/_analyze` 🟠 = partially tested in `/_snapshot/{repository}/_analyze` 🔴 = not tested in `/_snapshot/{repository}/_analyze` 🔵 = not testable in `/_snapshot/{repository}/_analyze`

Reviewer: nit Lets split each to a different line

**Why:** During snapshot operations, metadata blobs may be overwritten (e.g., updating the repository index file). Concurrent readers performing restore or listing snapshots must see a complete version of the blob, never a mixture of old and new content.

---

### **🟢 Aborted Writes Must Not Leave Visible Data**

**Description:** A write that fails or is aborted mid-stream must not leave a partial or complete blob visible to readers.

**Why:** If a node crashes or a write operation is cancelled mid-stream, the partial upload must not be visible. A partially-written metadata blob could corrupt the repository, and a partially-written data blob would cause checksum failures on restore. In stateless Elasticsearch, ObjectStoreService upload tasks explicitly track success to prevent accepting failures as success — a failed commit upload that leaves a visible blob would corrupt shard state.

---

### **🟢 Atomic Writes (No Partial Reads)**

**Description:** A blob must be either fully visible or not visible at all. Readers must never see a partially-written blob.

> **Why:** Snapshot metadata and data blobs must be fully written before they are visible to readers. A partially-written index file or snapshot metadata blob would corrupt the repository state and could cause data loss during restore. In stateless Elasticsearch, ObjectStoreService uses writeBlobAtomic for batched compound commits (merged Lucene index files up to ~15 MB) and writeBlob for translog files — partial visibility of a commit would cause shard recovery to read corrupted Lucene data.

Reviewer: BCCs can go over 15MB. A single huge commit of 100MB or even GBs is something we meet often. So lets skip the parenthesis.

Author: I don't think I follow. You're saying cut the part about `No Partial Reads`?

Reviewer: @masseyke No. I'm saying to remove the `up to ~15MB` part because it is incorrect.

---

## **Registers**

### **🟢 Linearizable Compare-And-Exchange (Register Operations)**

**Description:** The object store must provide linearizable semantics — operations appear to take effect at a single instant. This can be done by supporting one of two options: (1) conditional writes (preferred) or (2) multipart upload consensus protocol (fallback). See below for more information.

**Why:** Register operations are the foundation of distributed coordination across Elasticsearch. For snapshot repositories, multiple nodes may concurrently try to update shared state (e.g., the current generation of a repository). For stateless Elasticsearch, the CAS register is the mechanism for master election and cluster lease management — StatelessElectionStrategy uses compareAndSetRegister to claim terms and update leases, and StatelessClusterConsistencyService continuously verifies local cluster state against the lease stored in the object store via getRegister. Without linearizable CAS, split-brain scenarios or duplicate masters are possible.

**Implementation:** The S3 blob container provides two implementation paths for this requirement. A backend must support at least one:

> * **Path 1 — Conditional writes (preferred):** Uses `If-Match: <etag>` and `If-None-Match: *` headers on PutObject and CompleteMultipartUpload to implement CAS directly. This is the simpler, more efficient path. Conditional write support is gated by the `s3.client.<name>.unsafely_incompatible_with_s3_conditional_writes` setting — when false, conditional writes are enabled and this path is used. This path requires:

Reviewer: ```suggestion
* **Path 1 — Conditional writes (preferred):** Uses `If-Match: <etag>` and `If-None-Match: *` headers on PutObject and CompleteMultipartUpload to implement CAS directly. This is the simpler, more efficient path. Conditional write support is gated by the `unsafely_incompatible_with_s3_conditional_writes` S3 repository setting — when false (the default), conditional writes are enabled and this path is used. This path requires:
```

    1. 🟠 **ETag correctness:** GetObject and CompleteMultipartUpload responses must return stable, correct ETags. The CAS mechanism reads a blob's ETag, then writes back with `If-Match: <etag>` to ensure no concurrent modification. If ETags are missing, unstable, or incorrect, CAS breaks. *(Implicitly tested — CAS would fail without correct ETags — but not independently verified.)*
    2. 🟠 **Conditional header semantics:** `If-None-Match: *` must reject writes to existing blobs. `If-Match: <etag>` must reject writes when the current ETag doesn't match. Condition failures must return HTTP 409 or 412. *(Implicitly tested — CAS would fail without correct accept/reject behavior — but not independently verified at the HTTP level.)*
    3. 🟢 **failIfAlreadyExists protection:** When conditional writes are available, regular blob writes (writeBlob, writeBlobAtomic with failIfAlreadyExists=true) use `If-None-Match: *` to prevent concurrent writers from silently overwriting each other's data. This protection is lost when conditional writes are unavailable.

> * **Path 2 — Multipart upload consensus protocol (fallback):** When conditional writes are unavailable (i.e., `unsafely_incompatible_with_s3_conditional_writes` is true, the default), the implementation uses a multi-step consensus protocol to achieve CAS semantics. Note: this path does NOT provide failIfAlreadyExists protection on regular blob writes — the setting is named "unsafely" for this reason. This path requires, beyond the basic multipart lifecycle (Writes section, Multipart Upload Lifecycle):

Reviewer: ```suggestion
* **Path 2 — Multipart upload consensus protocol (fallback):** When conditional writes are unavailable (i.e., `unsafely_incompatible_with_s3_conditional_writes` is true), the implementation uses a multi-step consensus protocol to achieve CAS semantics. Note: this path does NOT provide failIfAlreadyExists protection on regular blob writes — the setting is named "unsafely" for this reason. This path requires, beyond the basic multipart lifecycle (Writes section, Multipart Upload Lifecycle):
```
